common/formatting/layout_optimizer.cc

```diff
@@ -465,13 +484,123 @@ void TokenPartitionsLayoutOptimizer::Optimize(int indentation,
+        node.Value().TokensRange().size() < kWrapTokensLimit) {
+      return factory_.WrappedLine(node.Value());
+    } else {
+      return factory_.Line(node.Value());
+    }
+  }
+
+  // Traverse and calculate children layouts
+
+  absl::FixedArray<LayoutFunction> layouts(node.Children().size());
+
+  switch (node.Value().PartitionPolicy()) {
+    case PartitionPolicyEnum::kJuxtaposition:
+    case PartitionPolicyEnum::kAlreadyFormatted:
+    case PartitionPolicyEnum::kWrap:
+    case PartitionPolicyEnum::kFitOnLineElseExpand:
+    case PartitionPolicyEnum::kAppendFittingSubPartitions:
+    case PartitionPolicyEnum::kJuxtapositionOrIndentedStack: {
+      std::transform(node.Children().begin(), node.Children().end(),
+                     layouts.begin(), [=](const TokenPartitionTree& n) {
+                       return this->CalculateOptimalLayout(n);
+                     });
+      break;
+    }
 
-  const auto calculate_optimal_layout_func =
-      std::bind(&TokenPartitionsLayoutOptimizer::CalculateOptimalLayout, this,
-                std::placeholders::_1);
+    case PartitionPolicyEnum::kStack:
+    case PartitionPolicyEnum::kAlwaysExpand:
+    case PartitionPolicyEnum::kTabularAlignment: {
+      int indentation = node.Value().IndentationSpaces();
+      std::transform(node.Children().begin(), node.Children().end(),
+                     layouts.begin(), [=](const TokenPartitionTree& n) {
+                       int relative_indentation =
+                           n.Value().IndentationSpaces() - indentation;
+                       if (relative_indentation < 0) {
+                         VLOG(0)
+                             << "(Child indentation) < (parent indentation). "
+                                "Assuming 0. Parent node:\n"
+                             << node;
+                       }
+
+                       LayoutFunction lf;
+
+                       if (relative_indentation > 0)
+                         lf = factory_.Indent(this->CalculateOptimalLayout(n),
+                                              relative_indentation);
+                       else
+                         lf = this->CalculateOptimalLayout(n);
+
+                       return lf;
+                     });
+      break;
+    }
+
+    case PartitionPolicyEnum::kUninitialized:
+    case PartitionPolicyEnum::kInline:
+      // Shouldn't happen - see the switch below.
+      break;
+  }
+
+  // Calculate and return current layout
 
   switch (node.Value().PartitionPolicy()) {
+    case PartitionPolicyEnum::kJuxtaposition:
+      return factory_.Juxtaposition(layouts.begin(), layouts.end());
+    case PartitionPolicyEnum::kStack:
+      return factory_.Stack(layouts.begin(), layouts.end());
+    case PartitionPolicyEnum::kWrap: {
+      if (VLOG_IS_ON(0) && node.Children().size() > 2) {
+        int second_indentation = node.Children()[1].Value().IndentationSpaces();
+        for (const auto& child : iterator_range(node.Children().begin() + 2,
```

Comment:
I've already added deduction guide for iterator_range: https://github.com/chipsalliance/verible/pull/1172/commits/8d1131176d5eb57c73b74b2fd475dd8b26618cde
Since Verible explicitly requires C++17 should we still care about older standards here?

Reply:
CTAD was be known to be buggy in various compilers leading up to 17, in particular implicit deduction, but I think these days we can rely on it to be working if we have a deduction guide.